graphql/query_resolver.go

```diff
@@ -963,6 +963,65 @@ func (r *queryResolver) TaskNamesForBuildVariant(ctx context.Context, projectIde
 	return buildVariantTasks, nil
 }
 
+// Waterfall is the resolver for the waterfall field.
+func (r *queryResolver) Waterfall(ctx context.Context, options WaterfallOptions) (*Waterfall, error) {
+	projectId, err := model.GetIdForProject(options.ProjectIdentifier)
+	if err != nil {
+		return nil, ResourceNotFound.Send(ctx, fmt.Sprintf("finding project with id: '%s'", options.ProjectIdentifier))
+	}
+	limit := model.DefaultWaterfallVersionLimit
+	if utility.FromIntPtr(options.Limit) != 0 {
+		if utility.FromIntPtr(options.Limit) > model.MaxWaterfallVersionLimit {
+			return nil, InputValidationError.Send(ctx, fmt.Sprintf("limit exceeds max limit of %d", model.MaxWaterfallVersionLimit))
+		}
+
+		limit = utility.FromIntPtr(options.Limit)
```

Comment:
nit: if this is going to reuse `utility.FromIntPtr(options.Limit)`, might be neater to put it in a var defined in the if like
```go
if limitOpt := utility.FromIntPtr(options.Limit); limitOpt != 0 {
    // ...
}```